### PR TITLE
feat: add semantic_search module (dense vector search)

### LINKED ISSUE #170: feat: add semantic_search module (dense vector search with Flat/IVF/LSH)
## Summary

Add a `semantic_search` module — the dense vector search complement to `sparse_search`. Single-file, stdlib-only, designed for LLM/Agent/RAG pipelines.

The module provides vector indexing and similarity search over pre-computed embeddings, with a pluggable `EmbeddingFunction` protocol and built-in stdlib API helpers (via `urllib.request`) for embedding providers supported by [llm-rosetta](https://github.com/Oaklight/llm-rosetta): **OpenAI**, **Cohere**, **Voyage**, **Jina**.

### Index Types (v1)

| Index | Algorithm | Build | Query at 100K | Recall | Use case |
|-------|-----------|-------|---------------|--------|----------|
| **Flat** | Brute force | O(1) | ~50ms (pure Python) | 100% | ≤5-10K vectors, exact baseline |
| **IVF** | K-means + inverted lists | O(n·k·d·iters) | ~5-15ms | ~93-97% | 10K-100K, best bang-for-buck in pure Python |
| **LSH** | Random projection hashing | O(n·L·K) | ~10-30ms | ~80-90% | Lightweight alternative, fast build |

HNSW intentionally excluded from v1 — in pure Python, build time is 5-20 min at 100K vectors for only marginal query-time improvement over IVF. PLAID (multi-vector/ColBERT-style) deferred to v2.

### Distance Metrics

- **Cosine** (default) — via pre-normalized vectors + dot product
- **Euclidean** (L2)
- **Inner Product**

Convention: smaller = more similar (negate IP, use `1 - cosine_sim`), consistent with Milvus.

### Embedding Integration

Three tiers, all stdlib-only:

1. **Pre-computed vectors** (always available):
   ```python
   index.add("doc1", vector=[0.1, 0.2, ...])
   ```

2. **Pluggable `EmbeddingFunction` protocol** (ChromaDB-style):
   ```python
   class EmbeddingFunction(Protocol):
       def __call__(self, texts: list[str]) -> list[list[float]]: ...
   
   index = SemanticIndex(embed=my_fn)
   index.add("doc1", text="hello world")  # auto-embeds
   ```

3. **Built-in API helpers** for llm-rosetta providers (using `urllib.request` + `json`):
   ```python
   from semantic_search import openai_embed, cohere_embed, voyage_embed, jina_embed
   embed = openai_embed(api_key="sk-...", model="text-embedding-3-small")
   index = SemanticIndex(embed=embed)
   ```

### Storage & Serialization

- Vectors: `array.array('f')` for memory efficiency (4 bytes/float vs 28+ in Python list)
- Persistence: binary via `struct` for vectors + JSON for metadata/config, auto-detected on load (same pattern as `sparse_search`)
- SQLite option for larger indices
- Optional **SQ8 quantization**: float32→uint8 per-dimension, 4x memory reduction, ~1% recall drop

### API Surface (following `sparse_search` conventions)

```python
__all__ = ["Result", "SemanticIndex", "rrf", "mmr",
           "openai_embed", "cohere_embed", "voyage_embed", "jina_embed"]

@dataclass
class Result:
    doc_id: str
    score: float
    metadata: dict[str, Any] | None = None

class SemanticIndex:
    def __init__(self, dim, metric="cosine", index_type="flat",
                 embed=None, quantize=False, **index_params)
    def add(self, doc_id, *, vector=None, text=None, metadata=None)
    def remove(self, doc_id)
    def update(self, doc_id, *, vector=None, text=None, metadata=None)
    def search(self, *, vector=None, text=None, top_k=10, filters=None) -> list[Result]
    def build_index(self)  # for IVF/LSH: explicit index construction
    def save(self, path, format=None)
    @classmethod
    def load(cls, path) -> SemanticIndex
```

`rrf` and `mmr` are reusable from `sparse_search` (or re-exported); enables hybrid sparse+dense fusion.

### Module Metadata

- **Tier**: `subsystem`
- **Category**: `text`
- **Deps**: `[]` (standalone, but `rrf`/`mmr` could optionally depend on `sparse_search` — TBD, may duplicate the functions instead)

### Performance Expectations (pure Python, stdlib-only)

| Scale | Flat query | IVF query | Build (IVF) | Memory (128d, M=16) |
|-------|-----------|-----------|-------------|---------------------|
| 1K | ~0.5ms | N/A | N/A | <1 MB |
| 10K | ~5ms | ~2-5ms | ~10-30s | ~6 MB |
| 50K | ~25ms | ~5-10ms | ~1-3min | ~30 MB |
| 100K | ~50ms | ~5-15ms | ~2-5min | ~61 MB |

With SQ8: vector memory is 4x smaller.

### Reference Libraries for Benchmarks

- **FAISS** (`faiss-cpu`) — correctness comparison for Flat, IVF, recall measurement
- **hnswlib** — optional additional comparison
- Add `bench-semantic_search` extra to `pyproject.toml`

## Research Basis

This design was informed by studying:
- **Milvus Lite v2** — pure Python rebuild, LSM-tree segments, SQ8 quantization, "smaller = more similar" convention
- **NextPlaid / ColGREP** — PLAID 3-stage pipeline, MaxSim scoring (deferred to v2)
- **ChromaDB** — `EmbeddingFunction` protocol design
- **RBSP framework** — proof that stdlib-only HNSW is feasible (but impractical at scale)
- **ANN-Benchmarks** — concrete performance numbers at scale
- **llm-rosetta** — supported embedding providers (OpenAI, Cohere, Voyage, Jina)

## Tasks

- [ ] Scaffold module with `zerodep new semantic_search --category text --tier subsystem`
- [ ] Implement core: distance metrics, `Result`, `SemanticIndex` with Flat index
- [ ] Implement IVF index (k-means + inverted lists)
- [ ] Implement LSH index (random projection)
- [ ] Implement `EmbeddingFunction` protocol + API helpers (OpenAI, Cohere, Voyage, Jina)
- [ ] Implement SQ8 quantization (optional)
- [ ] Implement persistence (binary vectors + JSON/SQLite metadata)
- [ ] Write correctness tests (against FAISS)
- [ ] Write benchmarks
- [ ] Register in manifest, module index config, docs
- [ ] Update bilingual docs (docs_en + docs_zh)

### PR DESCRIPTION
**Pull request by Oaklight**
## Summary

- Add `semantic_search` module — zero-dependency dense vector similarity search engine
- Supports Flat (exact), IVF (k-means clustering), and LSH (random projection) index types
- Includes `EmbeddingFunction` protocol + stdlib API helpers for OpenAI, Cohere, Voyage, Jina
- SQ8 scalar quantization for 4x memory reduction
- `detect_zh_en()` language detection utility for two-model embedding routing
- Dual-format persistence (JSON + SQLite) with auto-detection, matching `sparse_search` conventions
- `rrf()` and `mmr()` re-ranking utilities for hybrid sparse+dense fusion
- `Result` dataclass compatible with `sparse_search` for seamless interop

Closes #170

## Design decisions

- **No HNSW**: build time is impractical in pure Python (5-20 min at 100K vectors) for marginal query-time gain over IVF
- **No PLAID**: multi-vector/ColBERT-style search deferred to v2
- **API helpers use `urllib.request`**: zero extra dependencies. `openai_embed(base_url=...)` also works with Ollama, llama-server, and llamafile
- **Vectors stored in `array.array('f')`**: 4 bytes/float vs 28+ in Python lists, with stride-based offset access
- **Cosine via pre-normalization**: vectors L2-normalized on insert, reducing cosine distance to `1 - dot(a, b)`

## Test plan

- [x] 87 correctness tests passing (distance metrics, CRUD, all index types, persistence, rrf/mmr, edge cases)
- [x] 8 benchmark tests (FAISS correctness comparison, IVF/LSH recall, SQ8 accuracy)
- [x] Flat search ordering matches FAISS IndexFlatL2 exactly
- [x] Pre-commit hooks pass (ruff, ruff-format, ty, complexipy)
- [ ] CI green

### CODE REVIEW

File: semantic_search/semantic_search.py [hunk 1]
```diff
@@ -0,0 +1,1793 @@
...
+        if self._index_type == "ivf":
+            impl = self._index_impl
+            if isinstance(impl, _IVFIndex):
+                impl.build(self._vectors, self._dim, self._dist_fn, n_docs, seed=42)
+        elif self._index_type == "lsh":
+            impl = self._index_impl
+            if isinstance(impl, _LSHIndex):
+                impl.build(self._vectors, self._dim, self._dist_fn, n_docs, seed=42)
+
+        if self._quantize:
```
[the remark this answers is not in the document]
Comment (Oaklight) [reply semantic_search/semantic_search.py:1207]: Fixed in 3a5dc47 — clarified docstring: `quantize=True` computes SQ8 data for compact persistence, search always uses float32 for accuracy. Wiring quantized-domain distance computation into all three index types is a larger change; opening a follow-up issue for that.

File: semantic_search/semantic_search.py [hunk 2]
```diff
@@ -0,0 +1,1793 @@
...
+
+        n_docs = len(self._doc_ids)
+        raw = self._index_impl.search(
+            query, self._vectors, self._dim, self._dist_fn, n_docs, top_k
+        )
+
+        results: list[Result] = []
+        for idx, dist in raw:
+            did = self._doc_ids[idx]
+            if filters and not self._match_filters(did, filters):
```
[the remark this answers is not in the document]
Comment (Oaklight) [reply semantic_search/semantic_search.py:1167]: Fixed in 3a5dc47 — now over-fetches `min(top_k * 4, n_docs)` candidates when filters are set. Also documented the ANN limitation (restrictive filters on non-probed clusters may still return < top_k) in the docstring.

File: semantic_search/semantic_search.py [hunk 3]
```diff
@@ -0,0 +1,1793 @@
...
+
+    Returns:
+        ``"zh"`` or ``"en"``.
+    """
+    cjk = 0
+    latin = 0
+    for ch in text:
+        cp = ord(ch)
+        if (0x3040 <= cp <= 0x309F) or (0x30A0 <= cp <= 0x30FF):
+            return "en"
```
[the remark this answers is not in the document]
Comment (Oaklight) [reply semantic_search/semantic_search.py]: Fixed in 3a5dc47 — kana and Hangul characters are now counted alongside Latin in the denominator instead of triggering an early return. Mixed text like `"中文テスト"` correctly classifies as `"zh"` now. Added a test for this case.

File: semantic_search/semantic_search.py [hunk 4]
```diff
@@ -0,0 +1,1793 @@
...
+
+    Returns:
+        Parsed JSON response.
+
+    Raises:
+        urllib.error.URLError: On network or HTTP errors.
+    """
+    data = json.dumps(payload).encode("utf-8")
+    req = urllib.request.Request(url, data=data, headers=headers, method="POST")
+    with urllib.request.urlopen(req) as resp:
```
[the remark this answers is not in the document]
Comment (Oaklight) [reply semantic_search/semantic_search.py]: Fixed in 3a5dc47 — added `timeout=30` to `urlopen`.

File: semantic_search/semantic_search.py [hunk 5]
```diff
@@ -0,0 +1,1793 @@
...
+            for idx, doc_id, metadata_json in rows:
+                instance._doc_ids.append(doc_id)
+                instance._doc_id_to_idx[doc_id] = idx
+                instance._metadata[doc_id] = (
+                    json.loads(metadata_json) if metadata_json is not None else None
+                )
+
+            all_bytes = b""
+            for (data,) in cur.execute("SELECT data FROM vectors ORDER BY idx"):
+                all_bytes += data
```
[the remark this answers is not in the document]
Comment (Oaklight) [reply semantic_search/semantic_search.py]: Fixed in 3a5dc47 — replaced with `b"".join(chunks)`.

File: semantic_search/semantic_search.py [hunk 6]
```diff
@@ -0,0 +1,1793 @@
...
+# ---------------------------------------------------------------------------
+
+
+class SemanticIndex:
+    """Dense vector search index for semantic similarity.
+
+    Args:
+        dim: Vector dimensionality.
+        metric: Distance metric. One of ``"cosine"``, ``"euclidean"``,
+            ``"inner_product"``.
```
[the remark this answers is not in the document]
Comment (Oaklight) [reply semantic_search/semantic_search.py:920]: Fixed in 3a5dc47 — added "This class is not thread-safe. External locking is required for concurrent access from multiple threads." to the class docstring.

File: semantic_search/semantic_search.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (khiet-phan-covergo) [reply semantic_search/semantic_search.py]: Fixed in f80364340.

Confirmed and fixed. The routed path returned early into `RouteToStatusChangeCommand` and so never reached the composed path's `ApprovePremiumAsync`, leaving the adjustment in Draft — not issued, not billed.

The fix could not go at the dispatcher: `ApprovePolicyPremiumCommandHandler`'s allow-list excludes `Approved` and the routed command accepts internally, so approving after it returns can never succeed. It now runs inside `PolicyV2Service.SuspendPolicyEndorsement`, between `AdjustPremium` and `AcceptEndorsementAsync`. A failed approval returns the error and does not accept the endorsement (Req 2.1/2.2).

Added `GIVEN_Premium_Supplied_WHEN_SuspendPolicyEndorsement_THEN_ApprovesPremiumBeforeAccept` (asserts the exact order policyUpdated -> adjustorPersisted -> premiumApproved -> approved) and a failed-approval test asserting `AcceptEndorsementAsync` is never called.

File: semantic_search/semantic_search.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (Oaklight) [reply semantic_search/semantic_search.py:1207]: Fixed in aef0e3a — assertion messages now match the actual thresholds (IVF: 0.5, LSH: 0.05).

**Comment by Oaklight** [comment]
Addressed the remaining non-blocking suggestions from review in aef0e3a:

**4. Benchmark recall thresholds** — assertion messages now match actual thresholds (IVF: `>= 0.5`, LSH: `>= 0.05`).

**5. Deduplicate embed helpers** — extracted `_openai_compat_embed()` shared by `openai_embed`, `voyage_embed`, and `jina_embed`. `cohere_embed` stays separate (different API contract).

**6. `format` shadows builtin** — renamed to `fmt` in `save()` and `_resolve_format()`, updated tests.

**7. Batch `add` method** — added `add_many()` with batch embedding support (single `embed()` call for all texts). 9 tests covering basic ops, metadata, texts, error cases, and search correctness.
